src/test/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionListTest.java

```diff
@@ -160,6 +163,32 @@ public class FlowExecutionListTest {
         });
     }
 
+    @Test public void stepExecutionIteratorDoesNotLeakBuildsWhenOneIsStuck() throws Throwable {
+        sessions.then(r -> {
+            var notStuck = r.createProject(WorkflowJob.class, "not-stuck");
+            notStuck.setDefinition(new CpsFlowDefinition("semaphore('wait')", true));
+            var notStuckBuild = notStuck.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", notStuckBuild);
+            WeakReference<Object> notStuckBuildRef = new WeakReference<>(notStuckBuild);
+            // Create a Pipeline that runs a long-lived task on its CpsVmExecutorService, causing it to get stuck.
+            var stuck = r.createProject(WorkflowJob.class, "stuck");
+            stuck.setDefinition(new CpsFlowDefinition("echo 'test message'; Thread.sleep(Integer.MAX_VALUE)", false));
+            var stuckBuild = stuck.scheduleBuild2(0).waitForStart();
+            r.waitForMessage("test message", stuckBuild);
+            Thread.sleep(1000); // We need Thread.sleep to be running in the CpsVmExecutorService.
+            // Make FlowExecutionList$StepExecutionIteratorImpl.applyAll submit a task to the CpsVmExecutorService
+            // for stuck #1 that will never complete, so the resulting future will never complete.
+            StepExecution.applyAll(e -> null);
+            // Let notStuckBuild complete and check that it can be GC'd.
+            SemaphoreStep.success("wait/1", null);
+            r.waitForCompletion(notStuckBuild);
+            notStuckBuild = null; // Clear out the local variable in this thread.
+            Jenkins.get().getQueue().clearLeftItems(); // We don't want to wait 5 minutes.
+            MemoryAssert.assertGC(notStuckBuildRef, true);
+            // TODO: Test cleanup hangs for 1 minute in CpsFlowExecution.suspendAll because the checkpoint task can't run.
```

Comment:
I'm not sure if there is a good way to interrupt the VM thread here so the test can complete quickly. We can probably make the Pipeline call a static method in the test class that checks for us to set some flag here before continuing or something like that, or create a custom synchronous step that does something similar.

Reply:
Improved in [2de670d](https://github.com/jenkinsci/workflow-api-plugin/pull/347/commits/2de670dda3e6bdfa207e6c2b5bbcb2b56129a239) so that the test only takes 5ish seconds.